src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java

```diff
@@ -479,17 +480,29 @@ BlobMetadata buildAndUpload(String filePath, List<List<ChannelData>> blobData)
         byte[] compressedChunkData = BlobBuilder.compress(filePath, chunkData);
 
         // Encrypt the compressed chunk data, the encryption key is derived using the key from
-        // server with the full blob path
+        // server with the full blob path.
+        // We need to maintain IV as a block counter for the whole file, even interleaved,
+        // to align with decryption on the Snowflake query path.
+        long iv = curDataSize / Constants.ENCRYPTION_ALGORITHM_BLOCK_SIZE_BYTES;
         byte[] encryptedCompressedChunkData =
-            Cryptor.encrypt(compressedChunkData, firstChannel.getEncryptionKey(), filePath);
+            Cryptor.encrypt(compressedChunkData, firstChannel.getEncryptionKey(), filePath, iv);
+
+        // Encryption adds padding to the ENCRYPTION_ALGORITHM_BLOCK_SIZE_BYTES
+        // to align with decryption on the Snowflake query path starting from this chunk offset.
+        // The padding does not have arrow data and not compressed.
+        // Hence, the actual chunk size is smaller by the padding size.
+        // The compression on the Snowflake query path needs the correct size of the compressed
+        // data.
+        int padding = encryptedCompressedChunkData.length - compressedChunkData.length;
+        int encryptedCompressedChunkDataSize = encryptedCompressedChunkData.length;
+        int chunkLength = encryptedCompressedChunkDataSize - padding;
```

Comment:
i'm a little confused, so `chunkLength` should just be `compressedChunkData.length`?

Reply:
And could you add a log for `chunkLength` at line 527?

Reply:
good point, leftover